pkg/operator/ceph/cluster/mon/mon.go

```
@@ -365,42 +436,92 @@ func (c *Cluster) initClusterInfo(cephVersion cephver.CephVersion) error {
+		}
+		mons = append(mons, c.newMonConfig(c.maxMonID, zone))
+	}
+
+	return existingCount, mons, nil
+}
+
+func (c *Cluster) clusterInfoToMonConfig(excludedMon string) []*monConfig {
+	mons := []*monConfig{}
 	for _, monitor := range c.ClusterInfo.Monitors {
+		if monitor.Name == excludedMon {
+			// Skip a mon if it is being failed over
+			continue
+		}
+		var zone string
+		schedule := c.mapping.Schedule[monitor.Name]
+		if schedule != nil {
+			zone = schedule.Zone
+		}
 		mons = append(mons, &monConfig{
 			ResourceName: resourceName(monitor.Name),
 			DaemonName:   monitor.Name,
 			Port:         cephutil.GetPortFromEndpoint(monitor.Endpoint),
+			Zone:         zone,
 			DataPathMap: config.NewStatefulDaemonDataPathMap(
 				c.spec.DataDirHostPath, dataDirRelativeHostPath(monitor.Name), config.MonType, monitor.Name, c.Namespace),
 		})
 	}
-
-	// initialize mon info if we don't have enough mons (at first startup)
-	existingCount := len(c.ClusterInfo.Monitors)
-	for i := len(c.ClusterInfo.Monitors); i < size; i++ {
-		c.maxMonID++
-		mons = append(mons, c.newMonConfig(c.maxMonID))
-	}
-
-	return existingCount, mons
+	return mons
 }
 
-func (c *Cluster) newMonConfig(monID int) *monConfig {
+func (c *Cluster) newMonConfig(monID int, zone string) *monConfig {
 	daemonName := k8sutil.IndexToName(monID)
 
 	return &monConfig{
 		ResourceName: resourceName(daemonName),
 		DaemonName:   daemonName,
 		Port:         DefaultMsgr1Port,
+		Zone:         zone,
 		DataPathMap: config.NewStatefulDaemonDataPathMap(
 			c.spec.DataDirHostPath, dataDirRelativeHostPath(daemonName), config.MonType, daemonName, c.Namespace),
 	}
 }
 
+func (c *Cluster) findAvailableZoneIfStretched(mons []*monConfig) (string, error) {
+	if !c.spec.IsStretchCluster() {
+		return "", nil
+	}
+
+	// Build the count of current mons per zone
+	zoneCount := map[string]int{}
+	for _, m := range mons {
+		if m.Zone == "" {
+			logger.Errorf("zone not found on mon %q", m.DaemonName)
```

Isn't this validated by `preClusterStartValidation()`?

---

It's really a bug if this happened, I changed it to return an error instead of log